src/github.com/matrix-org/dendrite/syncapi/sync/userstream.go

```diff
@@ -15,65 +15,143 @@
 package sync
 
 import (
+	"context"
+	"runtime"
 	"sync"
+	"time"
 
 	"github.com/matrix-org/dendrite/syncapi/types"
+	"github.com/matrix-org/util"
 )
 
 // UserStream represents a communication mechanism between the /sync request goroutine
-// and the underlying sync server goroutines. Goroutines can Wait() for a stream position and
-// goroutines can Broadcast(streamPosition) to other goroutines.
+// and the underlying sync server goroutines.
+// Goroutines can get a UserStreamListener to wait for updates, and can Broadcast()
+// updates.
 type UserStream struct {
 	UserID string
-	// Because this is a Cond, we can notify all waiting goroutines so this works
-	// across devices for the same user. Protects pos.
-	cond *sync.Cond
-	// The position to broadcast to callers of Wait().
+	// The lock that protects changes to this struct
+	lock sync.Mutex
+	// Closed when there is an update.
+	signalChannel chan struct{}
+	// The last stream position that there may have been an update for the suser
 	pos types.StreamPosition
-	// The number of goroutines blocked on Wait() - used for testing and metrics
-	numWaiting int
+	// The last time when we had some listeners waiting
+	timeOfLastChannel time.Time
+	// The number of listeners waiting
+	numWaiting uint
+}
+
+// UserStreamListener allows a sync request to wait for updates for a user.
+type UserStreamListener struct {
+	*UserStream
```

Comment:
could we make this an explicit field? I find the implicit references to the UserStream members confusing.